man/man1/ismir_query.1.ronn
ismir_query(1) -- A tool to look for an audio query file in an audio index.
=============================================

## SYNOPSIS
Usage: ismir_query [SETTINGS]

## DESCRIPTION
ismir_query takes an audio query file (a wav file) and looks for it in the audio index.

## OPTIONS
  -q [ --query ] arg           Audio query file
  -d [ --dbname ] arg          Database name
  -p [ --processed-query ] arg Preprocessed query file

## AUTHOR
Jesper Sindahl Nielsen <jsn@statsbiblioteket.dk>


changelog
scape-xcorrsound (2.0.0) trusty; urgency=low
  * Fixed the debian packaging
  * Scape micro site
  * Install Instructions and version refs updated
  * README updates to match SCAPE template - work in 
  progress
  * Added descriptions for installing on Mac OS X
  * Tool dependencies as rpm packages added
  * Applied patch to remove the tpie dependency
  * Add command line flag --pad-short-block to control the 
  padding of samples shorter than the block length with 
  silence. Also add diagnostics to show more detail of 
  correlation failure if --verbose is given.
  * Correlation of short waveforms and Cygwin build
  * Pad short blocks read from the end of samples with 
  silence, so that the correlation is used in the case of 
  samples shorter than the block length (rather than exiting 
  the loop, leaving minimumVal as 2 and indicating success).
  * Additional includes necessary for building on Cygwin.
  * [Travis] test .deb and binaries
  * Added debian package to travis
  * And corrected project language yet again.
  * added a similarity measure output on success
  * Removed cout stuff
  * fixed a bug in AudioStream
  * received updates for tpie
  * build index now supports piping
  * quick and dirty wavStdin reader
  * now also compiling with TPIE and avcodec
  * prepared fingerprint_db for 'hashing-style' queries, 
  changed to using 64 samples skips, and various other things
  * added minor progress info for index building
  * files for CMake to find libavcodec
  * Got the revindex for testing purposes. Written by 
  Mathias Rav
  * changing design to have several fingerprinting 
  strategies
  * integrationTests now works with other paths
  * Removed the old debian specs as cmake does that job now
  * insertion and query works
  * ismir_build_index now working
  * Add travis file
  * fast comparisons
  * Added more information on why a waveform comparison 
  failed
  * Wrote up man pages for the tools
  * Added options: --threshold and --channel along with the 
  appropriate modifications for the parameters
  * added number of seconds pr block as a parameter to 
  waveform-compare
  * Renamed xcorrSound tool to the more descriptive 
  'overlap-analysis'
  * lets use a relational database
  * changed some things up. Fingerprints are now proper 
  size fx.
  * minor corrections and another way of indexing audio
  * fixed long long. Now int64_t
  * added return codes to migrationQA.cpp
  * A tool to help in creating the fingerprinting function 
  for a data set
  * sound index stuff. Various additions and changes. next 
  step is to do some test runs and see if the idea works.
  * Made a better silence detector for migrationQA
  * Lots of changes. Next step is to normalize the sound to 
  get more correct matches
  * Initial go at building an audio index
  * Fixed AudioStream to be more wav compliant. Now accepts 
  wav files with extra parameters
  * implemented the fingerprinting function. Need to fiddle 
  with the magic constant to get good results
  * migrationQA: now with negative offset and better info
  * migrationQA: tool now works. Remember input order 
  matters\!
  * soundMatch now works in a streaming fashion
 -- Jesper Sindahl Nielsen <jsn@statsbiblioteket.dk>  Wed, 23 Apr 2014 16:48:08 +0200

scape-xcorrsound (1.0.1) trusty; urgency=low
  * script for making the xcorrSound tarball
  * cleaning up fftw
  * sound match is now working
  * Fixed bugs in cross correlation
 -- Jesper Sindahl Nielsen <jsn@statsbiblioteket.dk>  Mon, 14 Nov 2011 13:54:48 +0000

scape-xcorrsound (1.0.0) trusty; urgency=low
  * Initial version
 -- Jesper Sindahl Nielsen <jsn@statsbiblioteket.dk>  Mon, 07 Nov 2011 14:34:30 +0000
 


apps/ismir_query.cc
#include <boost/program_options.hpp>
#include <fingerprint_db.hh>
#include <iostream>
#include <string>
#include <vector>

using std::vector;

namespace po = boost::program_options;
namespace si = sound_index;

void query(std::string &dbname, std::string &queryFile) {
    std::vector<std::string> res;
    si::fingerprint_db db;
    db.open(dbname);
    db.query_scan(queryFile, res);
    db.close();

    for (size_t i = 0; i < res.size(); ++i) {
        std::cout << res[i] << std::endl;
    }

}

void init(int argc, char *argv[]) {

    std::string dbname;
    std::string queryFile;
    std::string processedQueryFile;
    
    po::options_description generic("Program options");
    generic.add_options()
        ("help,h", "Print help message and return, everything else is ignored");
    
    po::options_description hidden("Settings");
    hidden.add_options()
	("query,q", po::value<std::string>(), "Audio query file")
	("dbname,d", po::value<std::string>(), "Database name")
	("processed-query,p", po::value<std::string>(), "Preprocessed query file");

    po::options_description all("Allowed options");
    all.add(generic).add(hidden);

    po::options_description visible("Program options");
    visible.add(generic);

    po::variables_map vm;
    po::store(po::command_line_parser(argc, argv).options(all).run(), vm);

    if (vm.count("help")) {
        std::cout << "Usage: ismir_query SETTINGS" << std::endl << std::endl;
        std::cout << generic << std::endl;

        std::cout << hidden << std::endl;

        exit(0);
    }

    bool queryExists = false;

    if (vm.count("query")) {
        queryFile = vm["query"].as<std::string>();
	queryExists = true;
    } else {
        std::cout << "A query file must be provided" << std::endl;
        exit(1);
    }

    if (vm.count("processed-query")) {
	processedQueryFile = vm["processed-query"].as<std::string>();
    } else {
	queryExists = true;
    }

    if (vm.count("dbname")) {
        dbname = vm["dbname"].as<std::string>();
    } else {
        std::cout << "dbname must be provided" << std::endl;
        exit(1);
    }

    query(dbname, queryFile);

}

int main(int argc, char *argv[]) {
    init(argc, argv);
    return 0;
}
